### PR TITLE
InstSimplify: support floating-point equivalences

### PR DESCRIPTION
**Pull request by artagnon**
Since cd16b07 (IR: introduce CmpInst::isEquivalence), there is now an isEquivalence routine in CmpInst that we can use to determine equivalence in simplifySelectWithICmpEq. Implement this, extending the code from integer-equalities to integer and floating-point equivalences.

### CODE REVIEW

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 1]
```diff
@@ -4635,33 +4635,29 @@ static Value *simplifySelectWithICmpEq(Value *CmpLHS, Value *CmpRHS,
 
 /// Try to simplify a select instruction when its condition operand is an
 /// integer comparison.
-static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
-                                         Value *FalseVal,
-                                         const SimplifyQuery &Q,
-                                         unsigned MaxRecurse) {
+static Value *simplifySelectWithCmpCond(Value *CondVal, Value *TrueVal,
```
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4638]: I don't think this function should change, nearly all code in here is not relevant for fcmp. Add a call to simplifySelectWithEquivalence in simplifySelectWithFCmp instead.

Comment (artagnon) [reply llvm/lib/Analysis/InstructionSimplify.cpp:4638]: Wouldn't there be code duplication then? What are your thoughts on absorbing simplifySelectWithFCmp into this function?

Comment (nikic) [reply llvm/lib/Analysis/InstructionSimplify.cpp:4638]: I don't think there is any substantial amount of overlap between the functions. Note that while this function contains a lot of calls to simplifySelectWithICmpEq, only the plain variant of them is relevant to the FP case. The ones involving and/or are integer-only.

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 2]
```diff
@@ -4761,28 +4761,54 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
...
   FCmpInst::Predicate Pred;
-  if (!match(Cond, m_FCmp(Pred, m_Specific(T), m_Specific(F))) &&
-      !match(Cond, m_FCmp(Pred, m_Specific(F), m_Specific(T))))
+  Value *CmpLHS, *CmpRHS;
+  if (!match(Cond, m_FCmp(Pred, m_Value(CmpLHS), m_Value(CmpRHS))))
+    return nullptr;
+  FCmpInst *I = cast<FCmpInst>(Cond);
+
+  bool IsEquiv = I->isEquivalence(),
+       IsInverseEquiv = I->isEquivalence(/*Invert=*/true);
```
Comment (arsenm) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4773]: This API feels awkward to check the inverse case. Maybe isEquivalence can return the predicate where it is equivalent if inverted?

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 3]
```diff
@@ -4761,28 +4761,54 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
...
       Q.CxtI && isa<FPMathOperator>(Q.CxtI) && Q.CxtI->hasNoSignedZeros();
-  const APFloat *C;
-  if (HasNoSignedZeros || (match(T, m_APFloat(C)) && C->isNonZero()) ||
-      (match(F, m_APFloat(C)) && C->isNonZero())) {
+  if (HasNoSignedZeros) {
     // (T == F) ? T : F --> F
-    // (F == T) ? T : F --> F
-    if (Pred == FCmpInst::FCMP_OEQ)
+    if (Pred == FCmpInst::FCMP_OEQ ||
+        (Pred == FCmpInst::FCMP_UEQ && I->hasNoNaNs()))
```
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4806]: Unrelated change?

**Review by nikic — CHANGES_REQUESTED** [review]
I think this is missing tests for the non-refining code path. Try something along the lines of folding `(x == 1) ? y : (x * y)` to `x * y`.

And related to that, I think we need to be extra careful about derefinement here. For integers, the only derefinement we really have to worry about is related to undef and poison, which the existing code already does. But for FP we also have to worry about non-determinism relating to libcalls and NaN behavior.

For example, if we have something like `x == C : sin(C) : sin(x)`, then we should generally not fold that to `sin(x)`, because we cannot guarantee that `sin(x)` at runtime folds to the same `sin(C)` that the host libm produced.

We should be able to avoid this by passing the non-determinism flag to the constant folding API in simplifyWithOpReplaced, for the non-refining case.

File: llvm/lib/Analysis/InstructionSimplify.cpp (hunk 2 above)
Comment (artagnon) [reply llvm/lib/Analysis/InstructionSimplify.cpp:4773]: I don't think there a way around this awkward API: if we return the predicate where it is equivalent, we still need to check three different predicates to tell whether or not to invert the TrueVal and FalseVal. I thought about returning an `std::optional<bool>` which would convey all the information with one call, but it's not much better than calling it twice and keeping two bools: we would be doing the same amount of work, and at least we wouldn't be uglifying isEquivalence.

**Comment by artagnon** [comment]
> For example, if we have something like `x == C : sin(C) : sin(x)`, then we should generally not fold that to `sin(x)`, because we cannot guarantee that `sin(x)` at runtime folds to the same `sin(C)` that the host libm produced.
> 
> We should be able to avoid this by passing the non-determinism flag to the constant folding API in simplifyWithOpReplaced, for the non-refining case.

Thanks for catching this subtle point. I do think `x == C : sin(x) : sin(C)` should fold, as I've demonstrated in the test case.

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 4]
```diff
@@ -4761,27 +4763,51 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
...
-  // This transform is safe if we do not have (do not care about) -0.0 or if
-  // at least one operand is known to not be -0.0. Otherwise, the select can
-  // change the sign of a zero operand.
+  // This transform is also safe if we do not have (do not care about) -0.0.
   bool HasNoSignedZeros =
       Q.CxtI && isa<FPMathOperator>(Q.CxtI) && Q.CxtI->hasNoSignedZeros();
-  const APFloat *C;
-  if (HasNoSignedZeros || (match(T, m_APFloat(C)) && C->isNonZero()) ||
-      (match(F, m_APFloat(C)) && C->isNonZero())) {
+  if (HasNoSignedZeros) {
```
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4805]: Why isn't this case already covered by the equivalence handling?

Comment (artagnon) [reply llvm/lib/Analysis/InstructionSimplify.cpp:4805]: Isn't the nsz on the select, which the isEquivalence can't possibly know? An nsz on the CmpInst doesn't tell us much.

Comment (artagnon) [reply llvm/lib/Analysis/InstructionSimplify.cpp:4805]: We could pass an IgnoreZeros flag to CmpInst::isEquivalence, and check HasNoSignedZeros to shorten this code, but I think that might not be worth the trouble and inelegance of an API taking two bools.

Comment (nikic) [reply llvm/lib/Analysis/InstructionSimplify.cpp:4805]: Oh, I missed that this is checking nsz on a different instruction.

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 5]
```diff
@@ -4761,27 +4763,51 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
...
+    if (Value *V =
+            simplifySelectWithEquivalence(CmpLHS, CmpRHS, T, F, Q, MaxRecurse))
+      return V;
+    if (Value *V =
+            simplifySelectWithEquivalence(CmpRHS, CmpLHS, T, F, Q, MaxRecurse))
+      return V;
+
+    // (T == F) ? T : F --> F
+    // (T != F) ? F : T --> F
+    return CmpLHS == T && CmpRHS == F ? F : nullptr;
```
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4796]: But shouldn't this case at least be covered by the preceding simplify code?

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 6]
```diff
@@ -4761,27 +4763,47 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
...
+
+  bool IsEquiv = I->isEquivalence(),
+       IsInverseEquiv = I->isEquivalence(/*Invert=*/true);
+
+  if (IsInverseEquiv)
+    std::swap(T, F);
+
+  // Canonicalize CmpLHS to be T, and CmpRHS to be F, if they're swapped.
+  if (CmpLHS == F && CmpRHS == T)
+    std::swap(CmpLHS, CmpRHS);
```
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4782]: I'd move this below the IsEquiv block, because it's only relevant for the code after it.

File: llvm/lib/Analysis/InstructionSimplify.cpp [hunk 7]
```diff
@@ -4761,27 +4763,47 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
...
+  Value *CmpLHS, *CmpRHS;
+  if (!match(Cond, m_FCmp(Pred, m_Value(CmpLHS), m_Value(CmpRHS))))
+    return nullptr;
+  FCmpInst *I = cast<FCmpInst>(Cond);
+
+  bool IsEquiv = I->isEquivalence(),
+       IsInverseEquiv = I->isEquivalence(/*Invert=*/true);
+
+  if (IsInverseEquiv)
+    std::swap(T, F);
```
Comment (nikic) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4778]: This is leaving behind a big footgun. I think the code here should be:

```
if (I->isEquivalence(/*Invert=*/true)) {
  std::swap(T, F);
  Pred = FCmpInst::getInversePredicate(Pred);
  IsEquiv = true;
}
```
so the values are correct after the swap.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by nikic** [comment]
Alive reports a verification failure for one of the tests: https://alive2.llvm.org/ce/z/eHm-rS

```llvm
----------------------------------------
define half @src(half %x, half %y) {
#0:
  %fcmp = fcmp oeq half %x, 0x3c00
  %fmul = fmul half %y, %x
  %sel = select i1 %fcmp, half %y, half %fmul
  ret half %sel
}
=>
define half @tgt(half %x, half %y) {
#0:
  %fmul = fmul half %y, %x
  ret half %fmul
}
Transformation doesn't verify!

ERROR: Value mismatch

Example:
half %x = #x3c00 (1)
half %y = #xfc80 (SNaN)

Source:
i1 %fcmp = #x1 (1)
half %fmul = #x7e00 (QNaN)
half %sel = #xfc80 (SNaN)

Target:
half %fmul = #x7c80 (SNaN)
Source value: #xfc80 (SNaN)
Target value: #x7c80 (SNaN)
```

The problem is that the fmul may return a different NaN value than just passing the value through.

cc @nunoplopes

**Comment by nunoplopes** [comment]
Yep, the optimization needs the 'nnan' flag. @jcranmer-intel to confirm.

**Comment by nikic** [comment]
I was wondering why this gets folded at all, given that this is the AllowRefinement case. I think the reason is this code: https://github.com/llvm/llvm-project/blob/76f258920d3baf32be297f60bee5b8520f195c25/llvm/lib/Analysis/InstructionSimplify.cpp#L4347-L4352

We assume that folding `x op id` to `x` is non-refining, which is not strictly true for FP.
